lib_loader.go

```diff
@@ -0,0 +1,67 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Purego only works on linux/macOS with amd64 and arm64 from now
+//go:build (linux || darwin) && (amd64 || arm64) && !cgo
+
+package waf
+
+import (
+	"fmt"
+
+	"github.com/ebitengine/purego"
+)
+
+// libLoader is created to wraps all interactions with the purego library
+type libLoader struct {
+	handle uintptr
+}
+
+// dlOpen open a handle for the shared library `name`.
+// the libLoader object is only a wrapped type for the linker handle
+// The `symbolsNeeded` is the list of symbols needed in the lib to load.
+// These symbols are returned via the map of symbols
+func dlOpen(name string, symbolsNeeded []string) (*libLoader, map[string]uintptr, error) {
```

Comment:
I suggest we make this API even more dynamic thanks to a new field tag `dlsym` so that we can directly fill in a Go struct with the expected symbols:

```go
type wafSymbols struct {
	Ddwaf_ruleset_info_free  uintptr `dlsym:"ddwaf_ruleset_info_free"`
	Ddwaf_init               uintptr `dlsym:"ddwaf_init"`
	Ddwaf_object_free        uintptr `dlsym:"ddwaf_object_free"`
	Ddwaf_destroy            uintptr `dlsym:"ddwaf_destroy"`
	Ddwaf_required_addresses uintptr `dlsym:"ddwaf_required_addresses"`
	Ddwaf_get_version        uintptr `dlsym:"ddwaf_get_version"`
	Ddwaf_context_init       uintptr `dlsym:"ddwaf_context_init"`
	Ddwaf_context_destroy    uintptr `dlsym:"ddwaf_context_destroy"`
	Ddwaf_result_free        uintptr `dlsym:"ddwaf_result_free"`
	Ddwaf_run                uintptr `dlsym:"ddwaf_run"`
}
```

and this function signature would become

```go
func dlOpen(name string, symbolsNeeded interface{}) (*libLoader, error)
// usage:
var symbols wafSymbols
lib, err := dlOpen("libddwaf.so", &symbols)
```

I think this will simplify the implementation of `newWafLoader` with a single source of symbol names `wafSymbols` in my example instead of three different Go values (the final struct, the map of symbol addresses, and the list of symbol names).

Reply:
Excellent idea, I'm on it :+1:

Reply:
done